go/bind/keybase.go

```diff
@@ -391,21 +398,131 @@ func unboxNotification(ctx context.Context, strConvID, body string, intMembersTy
+	pusher.LocalNotification("failedpending",
+		"Heads up! One or more pending messages failed to send. Tap here to retry them.",
+		-1, "default", convID.String(), "chat.failedpending")
+}
+
 // AppWillExit is called reliably on iOS when the app is about to terminate
 // not as reliably on android
-func AppWillExit() {
+func AppWillExit(pusher PushNotifier) {
 	defer kbCtx.Trace("AppWillExit", func() error { return nil })()
-	kbCtx.AppState.Update(keybase1.AppState_BACKGROUNDFINAL)
+	ctx := context.Background()
+	convs, err := kbChatCtx.MessageDeliverer.ActiveDeliveries(ctx)
+	if err == nil && len(convs) > 0 {
+		// We are about to get killed with messages still to send, let the user know they will get
+		// stuck
+		pushPendingMessageFailure(convs[0], pusher)
+	}
+	kbCtx.AppState.Update(keybase1.AppState_BACKGROUND)
 }
 
 // AppDidEnterBackground notifies the service that the app is in the background
 // [iOS] returning true will request about ~3mins from iOS to continue execution
 func AppDidEnterBackground() bool {
 	defer kbCtx.Trace("AppDidEnterBackground", func() error { return nil })()
+	ctx := context.Background()
+	convs, err := kbChatCtx.MessageDeliverer.ActiveDeliveries(ctx)
+	if err != nil {
+		kbCtx.Log.Debug("AppDidEnterBackground: failed to get active deliveries: %s", err)
+		return false
+	}
+	if len(convs) > 0 {
+		kbCtx.Log.Debug("AppDidEnterBackground: active deliveries in play, setting background active")
+		kbCtx.AppState.Update(keybase1.AppState_BACKGROUNDACTIVE)
+		return true
+	}
 	SetAppStateBackground()
 	return false
 }
 
+func AppBeginBackgroundTaskNonblock(pusher PushNotifier) {
+	defer kbCtx.Trace("AppBeginBackgroundTaskNonblock", func() error { return nil })()
+	go AppBeginBackgroundTask(pusher)
+}
+
+// AppBeginBackgroundTask notifies us that an app background task has been started on our behalf. This
+// function will return once we no longer need any time in the background.
+func AppBeginBackgroundTask(pusher PushNotifier) (abort bool) {
+	defer kbCtx.Trace("AppBeginBackgroundTask", func() error { return nil })()
+	ctx := context.Background()
+	// Poll active deliveries in case we can shutdown early
+	beginTime := libkb.ForceWallClock(time.Now())
+	ticker := time.NewTicker(5 * time.Second)
+	appState := kbCtx.AppState.State()
+	if appState != keybase1.AppState_BACKGROUNDACTIVE {
+		kbCtx.Log.Debug("AppBeginBackgroundTask: not in background mode, early out")
+		return false
+	}
+	var g *errgroup.Group
+	g, ctx = errgroup.WithContext(ctx)
+	g.Go(func() error {
+		select {
+		case appState = <-kbCtx.AppState.NextUpdate(&appState):
+			kbCtx.Log.Debug(
+				"AppBeginBackgroundTask: app state change, aborting with no task shutdown: %v", appState)
+			abort = true
+			return errors.New("app state change")
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	})
+	g.Go(func() error {
+		for {
+			ch, cancel := kbChatCtx.MessageDeliverer.NextFailure()
+			defer cancel()
```

Comment:
i think it would make sense to remove this defer and call `cancel` before the return on line 483.  otherwise the `defer` stack will pile up with these funcs while the for loop runs...

Reply:
Sick point, will do